### ISSUE TITLE
`wrong_self_convention` does not fire

### ISSUE THREAD

**Issue by tcharding** [rust-lang/rust-clippy#9248]
### Summary

`wrong_self_convention` does not fire when clippy is run on code from the lint example.

### Lint Name

wrong_self_convention

### Reproducer

I tried this code:

```rust
#![deny(clippy::wrong_self_convention)]

/// Demo wrong_self_convention
#[allow(dead_code)]
pub struct Wrong {
    field: u32,
}

impl Wrong {
    /// Should trigger `clippy::wrong_self_convention`.
    pub fn as_str(self) -> &'static str {
        ""
    }
}

fn main() {
    let w = Wrong { field: 42 };
    let _ = w.as_str();
}
```

I expected to see this happen:

`clippy` to warn about the wrong use of self.

Instead, this happened:

No lint warnings.

### Version

```text
rustc 1.62.1 (e092d0b6b 2022-07-16)
binary: rustc
commit-hash: e092d0b6b43f2de967af0887873151bb1c0b18d3
commit-date: 2022-07-16
host: x86_64-unknown-linux-gnu
release: 1.62.1
LLVM version: 14.0.5
```

**Comment by alex-semenyuk** [comment]
The issue with `pub` at `pub struct Wrong {`. It works without `pub` [playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=26898b4974c04a7a03fb0afe6c43272f)

**Comment by tcharding** [comment]
Oh wow, thanks for the response. Do you have any idea _why_ the lint does not fire for public types? Public types are the ones I most want to be idiomatic.

**Comment by xFrednet** [comment]
We want to avoid linting code which can't be changed by the user. Clippy has a config value `avoid-breaking-exported-api` to avoid breaking items from the public API. You can set the value to `false` in `clippy.toml` to enable this and other lints on public items.

Ignoring public items is not always ideal, sadly it's hard to determine if an item can actually be seen outside the crate. (Unless `pub(crate)` is specified). So ignoring all public items is the best fix :)

**Comment by tcharding** [comment]
Awesome, thanks for the explanation. Closing as this is fully resolved now. Thanks again.
